python/ray/air/preprocessor.py

```diff
@@ -165,18 +166,62 @@ def _fit(self, dataset: Dataset) -> "Preprocessor":
     def _transform(self, dataset: Dataset) -> Dataset:
         # TODO(matt): Expose `batch_size` or similar configurability.
         # The default may be too small for some datasets and too large for others.
-        return dataset.map_batches(self._transform_pandas, batch_format="pandas")
+        dataset_format = dataset._dataset_format()
+        exc_msg = "Neither `_transform_arrow` nor `_transform_pandas` are implemented."
+
+        # If only _transform_arrow is implemented, will convert the data to arrow.
+        # If only _transform_pandas is implemented, will convert the data to pandas.
+        # If both are implemented, will pick the method corresponding to the format
+        # for best performance.
+        # Implementation is defined as overriding the method in a sub-class.
+        if dataset_format == "arrow":
+            if self.__class__._transform_arrow != Preprocessor._transform_arrow:
+                return dataset.map_batches(
+                    self._transform_arrow, batch_format="pyarrow"
+                )
+            elif self.__class__._transform_pandas != Preprocessor._transform_pandas:
+                return dataset.map_batches(
+                    self._transform_pandas, batch_format="pandas"
+                )
+            else:
+                raise NotImplementedError(exc_msg)
+        elif dataset_format == "pandas":
+            if self.__class__._transform_pandas != Preprocessor._transform_pandas:
+                return dataset.map_batches(
+                    self._transform_pandas, batch_format="pandas"
+                )
+            elif self.__class__._transform_arrow != Preprocessor._transform_arrow:
+                return dataset.map_batches(
+                    self._transform_arrow, batch_format="pyarrow"
+                )
+            else:
+                raise NotImplementedError(exc_msg)
+        else:
+            raise ValueError(
+                f"Unsupported Dataset format: '{dataset_format}'. Only 'pandas' and "
+                "'arrow' Dataset formats are supported."
+            )
```

Comment:
@ericl This is a great motivator for providing something like a `"zero_copy"` batch format for library developers, where the batch type is dynamic (non-deterministic) and zero-copy/zero-conversions is guaranteed (the old semantics for `"native"`). This dispatching logic could be done within the UDF, dispatching based on the batch type rather than before the `.map_batches()` call, which would end up being a lot cleaner. This method would reduce to:
```python
 def _transform(self, dataset: Dataset) -> Dataset:
    return dataset.map_batches(self._transform, batch_format="zero_copy")
```